integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -70,10 +70,21 @@ public fun <T> CoroutineScope.future(
  * When `this` `ListenableFuture` is [successfully cancelled][java.util.concurrent.Future.cancel],
  * it will cancel the returned `Deferred`.
  *
- * When the returned `Deferred` is [cancelled][Deferred.cancel()], it will try to propagate the
+ * When the returned `Deferred` is [cancelled][Deferred.cancel], it will try to propagate the
  * cancellation to `this` `ListenableFuture`. Propagation will succeed, barring a race with the
  * `ListenableFuture` completing normally. This is the only case in which the returned `Deferred`
  * will complete with a different outcome than `this` `ListenableFuture`.
+ *
+ * Note: Cancellation of the coroutine that calls [await][Deferred.await] on the returned
+ * deferred, **does not** cancel [Deferred] and original [ListenableFuture].
+ * To propagate cancellation from awaiting coroutine to the [ListenableFuture], use [ListenableFuture.await].
+ *
+ * ```
```

Comment:
Why is this? The cancellation should be propagated no matter the adaption.

I added [this](https://github.com/vadimsemenov/kotlinx.coroutines/blame/c4b4e1c0211818ed3d01bdf988e7749cbb95f910/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L384) deferred-as-LF-as-deferred test case last time around. Did I miss a case, or is one of the `await()` implementations not cancellation-correct?

Reply:
Cancellation of the Deferred will cancel original ListenableFuture, that's fine. However, cancellation of _the coroutine that calls `Deferred.await`_ doesn't cancel the Deferred.

This is kind of "by design" of the Deferred. Deferred represents a computation that can be `await`ed multiple times. So one can say that it's unfair to cancel the Deferred if a coroutine that `await`s on the Deferred is cancelled. Because other coroutines that `await` on the same Deferred may succeed. 

Here is a simple example: https://pl.kotl.in/DYcyjkku2
If you replace the `async` coroutine builder with `future`, future will be cancelled as expected. 
I think it's the main difference between `ListenableFuture.await` and `Deferred.await`. `ListenableFuture.await` is intended for one-shot future. `Deferred.await` assumes that the deferred can be reused.

You can also take a look at [awaitFutureWithCancel](https://github.com/Kotlin/kotlinx.coroutines/blob/master/integration/kotlinx-coroutines-guava/test/ListenableFutureTest.kt#L549) in tests.

Reply:
OK, I'm following now.

I recommend striking the paragraph, then. The behaviour of `Deferred` isn't what this is documenting, it's documenting the adapter's behaviour.

Describing the cancellation of a calling coroutine and `await()` is adjacent to what this method is doing, and it confused me. I might need to care about that behaviour, but I should be discovering it through reading about `await()`, not here.

Some examples might be warranted, but I'm inclined to include positive examples rather than this one - cancellation works here and there, possibly contrasted with ordinary `Deferred` cancellation. It's also fair to leave those to the unit tests.

Reply:
Removed the paragraph. I think `await()` documents its own implementation good enough.